File: scripts/build_hw_catalog.py
import os, json, unicodedata

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
HW_ROOT = os.path.join(ROOT, 'Zona-Hardware')

CATEGORY_MAP = {
  'cpu': 'procesadores',
  'gpu': 'tarjetas-graficas',
  'almacenamiento': 'almacenamiento',
  'monitores': 'monitores',
  'perifericos': 'perifericos',
  'Gabinetes': 'gabinetes',
  'fuente-de-poder': 'fuentes',
  'ram': 'memoria',
  'motherboard': 'motherboard',
}

IMG_EXTS = {'.jpg','.jpeg','.png','.webp','.avif','.gif'}

def list_subdirs(d):
  try:
    return [os.path.join(d, x) for x in os.listdir(d) if os.path.isdir(os.path.join(d,x))]
  except: return []

def read_first_txt(d):
  try:
    for f in os.listdir(d):
      if f.lower().endswith('.txt'):
        with open(os.path.join(d,f), 'r', encoding='utf-8', errors='ignore') as fh:
          return fh.read().strip()
  except: pass
  return ''

def first_image(d):
  try:
    for f in os.listdir(d):
      if os.path.splitext(f)[1].lower() in IMG_EXTS:
        return f
  except: pass
  return ''

def slugify(s):
  s = unicodedata.normalize('NFKD', str(s))
  s = ''.join(c for c in s if not unicodedata.combining(c))
  out = []
  prev_dash = False
  for ch in s:
    if ch.isalnum():
      out.append(ch.lower())
      prev_dash = False
    else:
      if not prev_dash:
        out.append('-')
        prev_dash = True
  slug = ''.join(out).strip('-')
  return slug[:60]

def rel_path(p):
  return os.path.relpath(p, ROOT).replace(os.sep, '/')

def collect_category(cat_dir, cat_key):
  items = []
  slug_cat = CATEGORY_MAP.get(cat_key, cat_key)
  first_level = list_subdirs(cat_dir)
  for d in first_level:
    subdirs = list_subdirs(d)
    if subdirs and os.path.basename(cat_dir) == 'monitores':
      for sub in subdirs:
        img = first_image(sub)
        txt = read_first_txt(sub)
        if not img and not txt:
          continue
        title = os.path.splitext(img)[0] if img else (txt.split('\n')[0] or os.path.basename(sub))
        items.append({
          'id': f"{slug_cat}-{slugify(title)}",
          'title': title,
          'category': slug_cat,
          'image': rel_path(os.path.join(sub, img)) if img else '',
          'desc': txt,
        })
    else:
      img = first_image(d)
      txt = read_first_txt(d)
      if not img and not txt:
        continue
      title = os.path.splitext(img)[0] if img else (txt.split('\n')[0] or os.path.basename(d))
      items.append({
        'id': f"{slug_cat}-{slugify(title)}",
        'title': title,
        'category': slug_cat,
        'image': rel_path(os.path.join(d, img)) if img else '',
        'desc': txt,
      })
  return items

def build():
  items = []
  for cat_dir_name in CATEGORY_MAP.keys():
    cat_dir = os.path.join(HW_ROOT, cat_dir_name)
    if not os.path.exists(cat_dir):
      continue
    items.extend(collect_category(cat_dir, cat_dir_name))
  data = { 'items': items }
  out_file = os.path.join(HW_ROOT, 'catalogo.json')
  with open(out_file, 'w', encoding='utf-8') as f:
    json.dump(data, f, ensure_ascii=False, indent=2)
  print(f"Escrito catálogo con {len(items)} items en {out_file}")

if __name__ == '__main__':
  build()

File: scripts/build_hw_catalog.js
const fs = require('fs');
const path = require('path');

const root = path.resolve(__dirname, '..');
const hwRoot = path.join(root, 'Zona-Hardware');

const CATEGORY_MAP = new Map([
  ['cpu', 'procesadores'],
  ['gpu', 'tarjetas-graficas'],
  ['almacenamiento', 'almacenamiento'],
  ['monitores', 'monitores'],
  ['perifericos', 'perifericos'],
  ['Gabinetes', 'gabinetes'],
  ['fuente-de-poder', 'fuentes'],
  ['ram', 'memoria'],
  ['motherboard', 'motherboard'],
]);

function listSubdirs(dir){
  try{
    return fs.readdirSync(dir, { withFileTypes: true })
      .filter(d => d.isDirectory())
      .map(d => path.join(dir, d.name));
  }catch(e){ return []; }
}

function readFirstTxt(dir){
  const txts = fs.readdirSync(dir).filter(f => f.toLowerCase().endsWith('.txt'));
  if(txts.length === 0) return '';
  try{ return fs.readFileSync(path.join(dir, txts[0]), 'utf8').trim(); }catch(e){ return ''; }
}

function firstImage(dir){
  const exts = ['.jpg','.jpeg','.png','.webp','.avif','.gif'];
  const files = fs.readdirSync(dir);
  const found = files.find(f => exts.includes(path.extname(f).toLowerCase()));
  return found ? found : '';
}

function slugify(str){
  return String(str)
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g,'')
    .replace(/[^a-zA-Z0-9]+/g,'-')
    .replace(/^-+|-+$/g,'')
    .toLowerCase();
}

function relPath(p){
  return path.relative(root, p).split(path.sep).join('/');
}

function collectCategory(catDir, catKey){
  const out = [];
  const slugCat = CATEGORY_MAP.get(catKey) || catKey;
  // Algunos tienen subcarpetas especiales (p.ej. monitores/monitores (...))
  // Tomamos todas las carpetas hijas y también carpetas hijas de segundo nivel si la primera capa es un agrupador.
  const firstLevel = listSubdirs(catDir);
  for(const dir of firstLevel){
    const subdirs = listSubdirs(dir);
    if(subdirs.length > 0 && path.basename(catDir) === 'monitores'){
      // Caso monitores: monitores/monitores (...)
      for(const sub of subdirs){
        const img = firstImage(sub);
        const txt = readFirstTxt(sub);
        if(!img && !txt) continue;
        const title = img ? path.basename(img, path.extname(img)) : (txt.split('\n')[0] || path.basename(sub));
        out.push({
          id: `${slugCat}-${slugify(title).slice(0,60)}`,
          title: title,
          category: slugCat,
          image: relPath(path.join(sub, img)),
          desc: txt
        });
      }
    } else {
      const img = firstImage(dir);
      const txt = readFirstTxt(dir);
      if(!img && !txt) continue;
      const title = img ? path.basename(img, path.extname(img)) : (txt.split('\n')[0] || path.basename(dir));
      out.push({
        id: `${slugCat}-${slugify(title).slice(0,60)}`,
        title: title,
        category: slugCat,
        image: relPath(path.join(dir, img)),
        desc: txt
      });
    }
  }
  return out;
}

function build(){
  const items = [];
  for(const [catDirName] of CATEGORY_MAP){
    const catDir = path.join(hwRoot, catDirName);
    if(!fs.existsSync(catDir)) continue;
    const collected = collectCategory(catDir, catDirName);
    items.push(...collected);
  }
  const json = { items };
  const outFile = path.join(hwRoot, 'catalogo.json');
  fs.writeFileSync(outFile, JSON.stringify(json, null, 2), 'utf8');
  console.log(`Escrito catálogo con ${items.length} items en ${outFile}`);
}

build();
